search/streets_matcher.cpp

```diff
@@ -0,0 +1,165 @@
+#include "search/streets_matcher.hpp"
+#include "search/features_filter.hpp"
+#include "search/house_numbers_matcher.hpp"
+#include "search/query_params.hpp"
+
+#include "indexer/search_string_utils.hpp"
+
+#include "base/logging.hpp"
+#include "base/stl_helpers.hpp"
+
+namespace search
+{
+namespace
+{
+bool LessByHash(StreetsMatcher::Prediction const & lhs, StreetsMatcher::Prediction const & rhs)
+{
+  if (lhs.m_hash != rhs.m_hash)
+    return lhs.m_hash < rhs.m_hash;
+
+  if (lhs.m_prob != rhs.m_prob)
+    return lhs.m_prob > rhs.m_prob;
+
+  if (lhs.GetNumTokens() != rhs.GetNumTokens())
+    return lhs.GetNumTokens() > rhs.GetNumTokens();
+
+  return lhs.m_startToken < rhs.m_startToken;
+}
+}  // namespace
+
+// static
+void StreetsMatcher::Go(BaseContext const & ctx, FeaturesFilter const & filter,
+                        QueryParams const & params, vector<Prediction> & predictions)
+{
+  size_t const kMaxNumPredictions = 3;
+  double const kTailProbability = 0.05;
+
+  predictions.clear();
+  FindStreets(ctx, filter, params, predictions);
+
+  if (predictions.empty())
+    return;
+
+  sort(predictions.begin(), predictions.end(), &LessByHash);
+  predictions.erase(
+      unique(predictions.begin(), predictions.end(), my::EqualsBy(&Prediction::m_hash)),
+      predictions.end());
+
+  sort(predictions.rbegin(), predictions.rend(), my::LessBy(&Prediction::m_prob));
+  while (predictions.size() > kMaxNumPredictions && predictions.back().m_prob < kTailProbability)
```

Comment:
Sorry? I didn't get it, what complexity savings you are talking about? This constant is used to prevent large number of improbable results, that's why it's called Max, not Min.